### PR TITLE
16.0 mig membership prorate variable period

### PR DESCRIPTION
**Pull request by edlopen**
Standard migration to 16.0

@moduon MT-1833

@yajo @EmilioPascual @Shide @rafaelbn review this PR when you can. Thank you!

### CODE REVIEW

**Review by yajo — APPROVED** [review]

**Comment by yajo** [comment]
@chienandalu wanna review this one?

**Review by EmilioPascual — APPROVED** [review]

**Comment by edlopen** [comment]
Can you check it functionally @fcvalgar? Thanks!

**Review by fcvalgar — COMMENTED** [review]
I think the calculation of the prorate is not correct because it does not include the current day and the final day. In the memberships both days are usually included.

image

image

image

**Comment by edlopen** [comment]
Hi @fcvalgar , I've reviewed the two day difference that you commented. It is kind of a bummer because it seems that is missing the first and the last day and it lead us to misunderstood of what the problem really is.

I'll try to not enter too much in the details but this is a matter of decimal precision in the quantities.

In your case you are prorrating for the last 236 days in the year. If the quantity of a year is 1, our prorrated quantity should be 0.646575(...) years but in the invoice line will store only 0.64 which is that value rounded to two decimals. That lead us to $24 difference. So a membership that cost $1 a day make us think that we are missing the starting and ending date but, sadly, not.

To sum up, the difference that you notice is a precission error in the prorrate calculation. Because this is a migration I would consider that this PR is ok to merge but the roadmap will be updated to cover this issue later on.

I have a couple of ideas of how make this run propelly but I feel that is out of the scope of this PR.

Also, what do you think @EmilioPascual ?

**Review by fcvalgar — APPROVED** [review]
Functional review,

LGTM, thank you @edlopen

**Comment by yajo** [comment]
I think it's worth it to record your discoveries in the module's roadmap/known issues.

**Comment by EmilioPascual** [comment]
Perhaps error is in the rounding in this line https://github.com/OCA/vertical-association/blob/16.0/membership_prorate/models/account_move_line.py#L37, but it's in another module. 
Anyway migration is ok.

**Comment by edlopen** [comment]
@rafaelbn this PR is ok to merge. We've all agreed on this migration to move on. Thank you!

**Review by rafaelbn — APPROVED** [review]
/ocabot migration patch

**Comment by rafaelbn** [comment]
@EmilioPascual @edlopen about this:

> Perhaps error is in the rounding in this line https://github.com/OCA/vertical-association/blob/16.0/membership_prorate/models/account_move_line.py#L37, but it's in another module.
Anyway migration is ok.

Please open an issue in githug and a subtas in our side. 

thank you! 😄 You are awesome!
